core/nuxeo-connection.js

```diff
@@ -18,11 +18,180 @@ import 'nuxeo/nuxeo.js';
 import { html } from '@polymer/polymer/lib/utils/html-tag.js';
 import './nuxeo-element.js';
 
-{
-  // A global map of clients with connectionId as key.
-  // This map is shared between all instances of nuxeo-connection.
-  const nxClients = {};
+// A global map of clients with connectionId as key.
+// This map is shared between all instances of nuxeo-connection.
+const connections = {};
+
+export class Connection {
+  constructor(opts) {
+    this.id = (opts && opts.id) || 'nx';
+    this.url = (opts && opts.url) || '';
+    this.repositoryName = (opts && opts.repositoryName) || null;
+    this.username = (opts && opts.username) || null;
+    this.password = (opts && opts.password) || null;
+    this.method = (opts && opts.method) || 'basic';
+    this.token = (opts && opts.token) || null;
+  }
+
+  connect() {
+    // Create the client if needed
+    if (this.client) {
+      // if this instance does not have any properties set properties from the client
+      if (!this.url && !this.username && !this.password && !this.repositoryName) {
+        this.url = this.client._baseURL;
+        this.username = this.client._username;
+        this.password = this.client._password;
+        this.repositoryName = this.client._baseOptions.repositoryName;
+      }
+      // if properties match the existing client use it
+      if (
+        this.client._baseURL === this.url &&
+        this.client._username === this.username &&
+        this.client._password === this.password &&
+        this.client._baseOptions.repositoryName === this.repositoryName
+      ) {
+        // return the stored connection request promise and chain _handleConnected to update instance properties
+        return this.client._promise;
+      }
+      // otherwise override the client with the new properties
+      this.client = null;
+    }
+
+    const options = {
+      baseURL: this.url,
+      schemas: ['*'],
+    };
+
+    if (this.method === 'basic') {
+      if (this.username) {
+        options.auth = {
+          method: 'basic',
+          username: this.username,
+          password: this.password,
+        };
+      }
+    } else {
+      if (this.method === 'token' && this.token) {
+        options.auth = {
+          method: 'token',
+          token: this.token,
+        };
+      }
+      options.headers = { 'X-No-Basic-Header': true };
+    }
+
+    if (this.repositoryName) {
+      options.repositoryName = this.repositoryName;
+    }
+
+    // eslint-disable-next-line no-undef
+    this.client = this.client || new Nuxeo(options);
+
+    // share the connect promise between all instances (one per client)
+    this.client._promise = this.client.connect();
+
+    return this.client._promise.catch((error) => {
+      if (error.response.status === 401) {
+        if (this.method === 'form') {
+          // store url fragment in cookie
+          document.cookie = `nuxeo.start.url.fragment=${window.location.hash.substring(1) || ''}; path=/`;
```

Comment:
Stupid question: what is the goal of storing the url fragment in the cookie?

Reply:
I'm assuming it's for a redirect after the user re-authenticates? We could make that clearer by improving previous comment.